app/domain/authentication/authn_iam/authenticator.rb

```diff
@@ -6,72 +6,74 @@ module Authentication
-            Errors::Authentication::AuthnIam::IdentityVerificationErrorCode.new(
-              response.code
-            )
-          )
-          false
-        end
       end
 
-      def iam_role_matches?(login, response_hash)
-        split_assumed_role = response_hash["GetCallerIdentityResponse"]["GetCallerIdentityResult"]["Arn"].split(":")
+      private
+
+      # Parses STS Caller Identity response for the Conjur host and resolves it against the provided login
+      def iam_role_matches?(login:, aws_arn:, aws_account:)
+        arn_parts = aws_arn.split(':')
 
         # removes the last 2 parts of login to be substituted by the info from getCallerIdentity
-        host_prefix = (login.split("/")[0..-3]).join("/")
-        aws_role_name = split_assumed_role[5].split("/")[1]
-        aws_account_id = response_hash["GetCallerIdentityResponse"]["GetCallerIdentityResult"]["Account"]
-        aws_user_id = response_hash["GetCallerIdentityResponse"]["GetCallerIdentityResult"]["UserId"]
-        host_to_match = "#{host_prefix}/#{aws_account_id}/#{aws_role_name}"
+        host_prefix = (login.split('/')[0..-3]).join('/')
 
-        Rails.logger.debug(
+        aws_role_name = arn_parts[5].split('/')[1]
+        host_to_match = "#{host_prefix}/#{aws_account}/#{aws_role_name}"
+
+        @logger.debug(
           LogMessages::Authentication::AuthnIam::AttemptToMatchHost.new(
-            aws_user_id,
+            login,
             host_to_match
           )
         )
 
         login.eql?(host_to_match)
       end
 
-      def aws_signed_url
-        return 'https://sts.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15'      
+      # Returns the STS endpoint called based on the provided hostname as a URI
+      def aws_signed_url(host)
+        URI("https://#{host}/?Action=GetCallerIdentity&Version=2011-06-15")
+      end
+
+      def extract_relevant_data(response)
+        {
+          arn: response.dig('GetCallerIdentityResponse', 'GetCallerIdentityResult', 'Arn'),
+          account: response.dig('GetCallerIdentityResponse', 'GetCallerIdentityResult', 'Account')
+        }
       end
 
+      # Calls to AWS STS endpoint using the provided authentication header
       def response_from_signed_request(aws_headers)
-        Rails.logger.debug(LogMessages::Authentication::AuthnIam::RetrieveIamIdentity.new)
         begin
-          RestClient.get(aws_signed_url, headers = aws_headers)
-        rescue RestClient::ExceptionWithResponse => e
-          Rails.logger.error(Errors::Authentication::AuthnIam::VerificationError.new(e.to_s))
-          raise Errors::Authentication::AuthnIam::InvalidAWSHeaders, e.to_s
+          response = @client.get_response(aws_signed_url(aws_headers['host']), aws_headers)
+          body =  Hash.from_xml(response.body)
+
+          return extract_relevant_data(body) if response.code.to_i == 200
+        rescue StandardError => e
+          raise(Errors::Authentication::AuthnIam::VerificationError.new(e))
```

Comment:
What's going on with this error raise? Seems like it will only happen when a non-StandardError gets raised from the above client request. Is there even any way for that to happen? And if it did, are we sure it would be because of invalid AWS headers? Maybe add a comment because this seems like an odd error handling pattern.

Reply:
This catch handles any sort of network exception (timeout, etc.).  I'll add a comment for clarification.

Reply:
Oh wow, `Net::HTTP` errors don't extend `StandardError` huh? TIL.

Reply:
I know! I learned that yesterday as well.  It's a big reason why folks us the HTTP wrapper gems.